parity-crypto/src/hmac/test.rs

```diff
@@ -46,3 +47,170 @@ fn simple_mac_and_verify() {
+		&hex!("dddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddddd"),
+		&hex!("
+			773ea91e36800e46854db8ebd09181a7
+			2959098b3ef8c122d9635514ced565fe"),
+		&hex!("
+			fa73b0089d56a284efb0f0756c890be9
+			b1b5dbdd8ee81a3655f83e33b2279d39
+			bf3e848279a722c806b485a47e67c807
+			b946a337bee8942674278859e13292fb")
+	);
+
+	// Test Case 4
+	check_test_vector(
+		&hex!("0102030405060708090a0b0c0d0e0f10111213141516171819"),
+		&hex!("
+			cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd
+			cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd
+			cdcdcdcdcdcdcdcdcdcdcdcdcdcdcdcd
+			cdcd"),
+		&hex!("
+			82558a389a443c0ea4cc819899f2083a
+			85f0faa3e578f8077a2e3ff46729665b"),
+		&hex!("
+			b0ba465637458c6990e5a8c5f61d4af7
+			e576d97ff94b872de76f8050361ee3db
+			a91ca5c11aa25eb4d679275cc5788063
+			a5f19741120c4f2de2adebeb10a298dd")
+	);
+
+	// Test Case 6
+	check_test_vector(
+		&hex!("
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaa"),
+		&hex!("
+			54657374205573696e67204c61726765
+			72205468616e20426c6f636b2d53697a
+			65204b6579202d2048617368204b6579
+			204669727374"),
+		&hex!("
+			60e431591ee0b67f0d8a26aacbf5b77f
+			8e0bc6213728c5140546040f0ee37f54"),
+		&hex!("
+			80b24263c7c1a3ebb71493c1dd7be8b4
+			9b46d1f41b4aeec1121b013783f8f352
+			6b56d037e05f2598bd0fd2215d6a1e52
+			95e64f73f63f0aec8b915a985d786598")
+	);
+
+	// Test Case 7
+	check_test_vector(
+		&hex!("
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
+			aaaaaa"),
+		&hex!("
+			54686973206973206120746573742075
+			73696e672061206c6172676572207468
+			616e20626c6f636b2d73697a65206b65
+			7920616e642061206c61726765722074
+			68616e20626c6f636b2d73697a652064
+			6174612e20546865206b6579206e6565
+			647320746f2062652068617368656420
+			6265666f7265206265696e6720757365
+			642062792074686520484d414320616c
+			676f726974686d2e"),
+		&hex!("
+			9b09ffa71b942fcb27635fbcd5b0e944
+			bfdc63644f0713938a7f51535c3a35e2"),
+		&hex!("
+			e37b6a775dc87dbaa4dfa9f96e5e3ffd
+			debd71f8867289865df5a32d20cdc944
+			b6022cac3c4982b10d5eeb55c3e4de15
+			134676fb6de0446065c97440fa8c6a58")
+	);
+}
+
+#[test]
+fn secrets_are_zeroed_on_drop() {
+	let ptr: *const KeyInner;
+	{
+		let secret = b"sikrit";
+		let signing_key = SigKey::sha256(secret);
+		ptr = &signing_key.0;
+		unsafe {
+			assert_eq!(*ptr, KeyInner::Sha256(DisposableBox::from_slice(b"sikrit")));
+		}
+	}
+	unsafe {
+		assert_eq!(*ptr, KeyInner::Sha256(DisposableBox::from_slice(&[0u8;6][..])));
```

Comment:
Actually, if I replace `[0u8, 6]` with `[1u8; 6]`, it still passes on my laptop. My guess is after "sikrit" `DisposableBox` deallocation, it reuses the space for the next allocation. So it's better to allocate upfront. Also `--release` may affect code generation.

Reply:
Something like this:
```rust
	let ptr: *const KeyInner;
	let zeros = KeyInner::Sha256(DisposableBox::from_slice(&[0u8; 6][..]));
	let expected = KeyInner::Sha256(DisposableBox::from_slice(b"sikrit"));
	{
		let secret = b"sikrit";
		let signing_key = SigKey::sha256(secret);
		ptr = &signing_key.0;
		unsafe {
			assert_eq!(*ptr, expected);
		}
	}
	unsafe {
		assert_eq!(*ptr, zeros);
	}
```